### PR TITLE
CBL-4519: Remove API_AVAILABLE annotation for macOS 10.12+ and iOS 10.0+

### PR DESCRIPTION
**Pull request by velicuvlad**
https://issues.couchbase.com/browse/CBL-4519

- removed API_AVAILABLE where version is less than the minimum deployment target

### CODE REVIEW

**Comment by velicuvlad** [comment]
blocked by https://github.com/couchbase/couchbase-lite-core/pull/1807 - will rebase when merged

File: Crypto/PublicKey+Apple.mm [hunk 1]
```diff
@@ -165,20 +165,17 @@ __unused static string describe(CFTypeRef ref) {
...
-                    error::_throw(error::CryptoError, "Couldn't get the data of a public key");
-                }
-                alloc_slice result(data);
-                CFRelease(data);
-                return result;
-            } else {
-               LogError(TLSLogDomain, "Couldn't get the data of a public key: Not supported by macOS < 10.12 and iOS < 10.0");
-                error::_throw(error::UnsupportedOperation, "Not supported by macOS < 10.12 and iOS < 10.0");
+            CFErrorRef error;
+            ++gC4ExpectExceptions;  // ignore internal C++ exceptions in Apple Security framework
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:169]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 2]
```diff
@@ -165,20 +165,17 @@ __unused static string describe(CFTypeRef ref) {
...
+            CFDataRef data = SecKeyCopyExternalRepresentation(_publicKeyRef, &error);
+            --gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:171]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 3]
```diff
@@ -190,39 +187,40 @@ virtual alloc_slice publicKeyDERData() override {
...
-                        (id)kSecAttrKeyClass:           (id)kSecAttrKeyClassPrivate,
-                        (id)kSecAttrApplicationLabel:   publicKeyHash
-                    };
-                    status = SecItemDelete((CFDictionaryRef)params);
-                    if (status != errSecSuccess && status != errSecInvalidItemRef && status != errSecItemNotFound)
-                        checkOSStatus(status, "SecItemDelete", "Couldn't remove a private key from the Keychain");
+            @autoreleasepool {
+                // Get public key hash from kSecAttrApplicationLabel attribute:
+                // See: https://developer.apple.com/documentation/security/ksecattrapplicationlabel
+                ++gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:193]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 4]
```diff
@@ -190,39 +187,40 @@ virtual alloc_slice publicKeyDERData() override {
...
+                NSDictionary* attrs = CFBridgingRelease(SecKeyCopyAttributes(_privateKeyRef));
+                --gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:195]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 5]
```diff
@@ -190,39 +187,40 @@ virtual alloc_slice publicKeyDERData() override {
...
-                LogError(TLSLogDomain, "Couldn't remove keys: Not supported by macOS < 10.12 and iOS < 10.0");
-                throwMbedTLSError(MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE);
+                
+                // Delete Public Key:
+                NSDictionary* params = @ {
+                    (id)kSecClass:                  (id)kSecClassKey,
+                    (id)kSecAttrKeyClass:           (id)kSecAttrKeyClassPublic,
+                    (id)kSecAttrApplicationLabel:   publicKeyHash
+                };
+                ++gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:207]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 6]
```diff
@@ -190,39 +187,40 @@ virtual alloc_slice publicKeyDERData() override {
...
+                OSStatus status = SecItemDelete((CFDictionaryRef)params);
+                --gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:209]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 7]
```diff
@@ -190,39 +187,40 @@ virtual alloc_slice publicKeyDERData() override {
...
+                if (status != errSecSuccess && status != errSecInvalidItemRef && status != errSecItemNotFound)
+                    checkOSStatus(status, "SecItemDelete", "Couldn't remove a public key from the Keychain");
+                
+                // Delete Private Key:
+                params = @ {
+                    (id)kSecClass:                  (id)kSecClassKey,
+                    (id)kSecAttrKeyClass:           (id)kSecAttrKeyClassPrivate,
+                    (id)kSecAttrApplicationLabel:   publicKeyHash
+                };
+                ++gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:219]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 8]
```diff
@@ -190,39 +187,40 @@ virtual alloc_slice publicKeyDERData() override {
...
+                status = SecItemDelete((CFDictionaryRef)params);
+                --gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:221]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 9]
```diff
@@ -332,114 +320,102 @@ virtual int _sign(int/*mbedtls_md_type_t*/ mbedDigestAlgorithm,
...
-                    warnCFError(error, "SecKeyCreateRandomKey");
-                    return nullptr;
-                }
-                publicKey = SecKeyCopyPublicKey(privateKey);
-
-            } else {
-                OSStatus err = SecKeyGeneratePair((CFDictionaryRef)params, &publicKey, &privateKey);
-                checkOSStatus(err, "SecKeyGeneratePair", "Couldn't create a private key");
+            CFErrorRef error;
+            ++gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:324]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 10]
```diff
@@ -332,114 +320,102 @@ virtual int _sign(int/*mbedtls_md_type_t*/ mbedDigestAlgorithm,
...
+            privateKey = SecKeyCreateRandomKey((CFDictionaryRef)params, &error);
+            --gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:326]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 11]
```diff
@@ -332,114 +320,102 @@ virtual int _sign(int/*mbedtls_md_type_t*/ mbedDigestAlgorithm,
...
+                (id)kSecClass:                  (id)kSecClassKey,
+                (id)kSecAttrKeyType:            (id)kSecAttrKeyTypeRSA,
+                (id)kSecAttrKeyClass:           (id)kSecAttrKeyClassPrivate,
+                (id)kSecAttrApplicationLabel:   publicKeyHash(publicKey),
+                (id)kSecReturnRef:              @YES
+            });
+            if (!privateKeyRef)
+                return nullptr;
+            
+            ++gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:398]: Remove. Test needs to do this instead.

File: Crypto/PublicKey+Apple.mm [hunk 12]
```diff
@@ -332,114 +320,102 @@ virtual int _sign(int/*mbedtls_md_type_t*/ mbedDigestAlgorithm,
...
+            auto publicKeyRef = SecKeyCopyPublicKey(privateKeyRef);
+            --gC4ExpectExceptions;
```
Comment (pasin) [inline Crypto/PublicKey+Apple.mm:400]: Remove. Test needs to do this instead.

**Review by pasin — CHANGES_REQUESTED** [review]

**Comment by jianminzhao** [comment]
Approve contingent on Pasin's comments about gC4ExpectExceptions;
API_AVAILABLE is slick. Must be something with objective-c compiler.

**Review by pasin — APPROVED** [review]

**Pull request merged**